### ISSUE TITLE
False positive for "no visible binding for global variable" check

### ISSUE THREAD

**Issue by TCABJ** [r-lib/lintr#890]
It seems like the "no visible binding for global variable" check is producing false positives in a very specific scenario.

If I try to use `$` subsetting both inside and outside functions in a script this check will treat the name after the `$` inside the function body as a global variable.

Here's the simplest runnable example:
```R
fun <- function() {
  df2 <- data.frame(y = "another string")
  df2$y
}

my_list <- list()
my_list$b <- "a string"
fun()
```

Saving the above R code to a file named `test.R` and running the linter gives the following results:
```
> lintr::lint("test.R")
/home/tcabj/gresb-scoring/test.R:3:7: warning: no visible binding for global variable ‘y’
  df2$y
```

I experimented a bit and found that any of the following changes makes the error go away:
1. Moving the code inside the function body outside it
2. Removing the line `my_list$b <- "a string"`
These two characteristics of the code snippet seem to be what causes the error to occur.

My environment details:
- lintr version 2.0.1
- R version 4.0.4
- WSL2  Ubuntu 20.04.2 LTS

**Comment by AshesITR** [comment]
This should be fixed in #709. Can you try the development version of lintr?

**Comment by TCABJ** [comment]
It is indeed fixed in the development version. Will make sure to check that first before posting next time. Thanks!
